openfe/setup/molecule.py

```diff
@@ -1,10 +1,44 @@
 # This code is part of OpenFE and is licensed under the MIT license.
 # For details, see https://github.com/OpenFreeEnergy/openfe
 
+import io
 from typing import TypeVar
 RDKitMol = TypeVar('RDKitMol')
 
+from rdkit import Chem
+
 from openfe.utils.molhashing import hashmol
+import openfe
+
+
+def _ensure_ofe_name(mol, name):
+    """
+    Ensure that rdkit mol carries the ofe-name and ofe-version tags.
+    """
+    rdkit_name = name  # override this if the property is defined
+    if name == "":
+        try:
+            rdkit_name = mol.GetProp("ofe-name")
+        except KeyError:
+            pass
+
+    if name != "" and rdkit_name != name:
+        raise RuntimeError(f"Internal inconsistency: 'name' has "
+                           f"value '{rdkit_name}' in the internal RDKit "
+                           f"representation; expected '{name}'")
```

Comment:
I'm not sure this should be an error, if I'd like to rename a Mol, how do I do this?  I think specifying name=something different is intentionally changing the name.

Reply:
I see the point on renaming. However, I also see this as an easy source of user error, and it would be nice if we did something to prevent that. In the newest commit, I'm issuing a warning if you rename something -- this should prevent users from accidentally creating a problem that is only evident much further down the line while also allowing explicit renaming.

I don't think that renaming should be a common activity. You can get weird behavior like

```python
m1 = Molecule(rdkitmol, "name1")
m2 = Molecule(rdkitmol, "name2")
m1.rdkit.GetProp("ofe-name")
# 'name2'
```

Actually, given that, maybe renaming shouldn't be an allowed activity?

Reply:
I think this actually highlights a different problem, we should be copying (i.e. taking ownership) the input to Molecule, so this wouldn't happen

Reply:
I don't think this code is unthinkable...

```python

mol = Chem.MolFromSomewhere()
conf = mol.GetConformer(0)

conf = pose1
m1 = Molecule(mol, name='pose1')

conf = pose2
m2 = Molecule(mol, name='pose2')
```

Reply:
Yeah, that was sort of my reason for raising an error. Something like that might be tried, and if it can't do what a user would expect, we should do something to prevent the user from doing it.

What we really need is for the RDKit molecules we wrap to be immutable; I couldn't figure out how to do that. Copying still allows issues like this:

```python
m1 = Molecule(mol)
mol.SetProp('property-needed-by-planner')
network = network_planner([m1] + other_mols)  # m1.rdkit doesn't know the property; unexpected behavior
```

Another option is for us to be indifferent to the `ofe-*` properties except on loading and storing. In the same way that I manually write the `ofe-version`, I could manually write `ofe-name`. This would mean that OpenFE should never use the `ofe-name` or `ofe-version` properties. They would be set on saving. Probably should hint that they are private in that case: `_ofe_name`, `_ofe_version`.

Reply:
Actually, IMO this argues for the standard way to creating molecules (from user perspective) to be from a `.from_rdkit` classmethod. I have different expectations between:

```python
m1 = Molecule(rdkit=mol)
assert m.rdkit is mol  # I expect this to be true

m2 = Molecule.from_rdkit(mol)
# I don't have the same expectation here; even less so if we call it `m2.as_rdkit()`
```

Reply:
Ok add the constructor changes onto #58 and we can get that change done so this PR makes sense

Reply:
> Ok add the constructor changes onto https://github.com/OpenFreeEnergy/openfe/pull/58 and we can get that change done so this PR makes sense

Done in [c0829bf](https://github.com/OpenFreeEnergy/openfe/pull/58/commits/c0829bfae678a6992d3fbb69743367ab5eab18e8)